src/ide/code_actions/fill_trait_members.rs

```diff
@@ -0,0 +1,278 @@
+        item_impl.trait_path(db).elements(db),
+        NotFoundItemType::Trait,
+    ) {
+        Ok(ResolvedConcreteItem::Trait(id)) => Some(id),
+        _ => None,
+    }
+}
+
+/// Generates declaration of a [`TraitConstantId`] containing its name, type
+/// and a placeholder value.
+/// Generics are substituted with concrete types according to a given [`SubstitutionRewriter`]
+fn constant_code(
+    db: &AnalysisDatabase,
+    id: TraitConstantId,
+    type_concretizer: &mut SubstitutionRewriter<'_>,
+) -> Option<String> {
+    let name = id.name(db);
+    let ty = type_concretizer.rewrite(db.trait_constant_type(id).ok()?).ok()?.format(db);
+
+    Some(format!("const {name}: {ty} = ();"))
+}
+
+/// Generates declaration of a [`TraitFunctionId`] containing its signature with parameters,
+/// panic indicator, implicits and default implementation if such exists.
+/// Generics are substituted with concrete types according to a given [`SubstitutionRewriter`].
+/// Returns None if function has a default implementation.
+fn function_code(
+    db: &AnalysisDatabase,
+    id: TraitFunctionId,
+    type_concretizer: &mut SubstitutionRewriter<'_>,
+) -> Option<String> {
+    // Do not complete functions that have default implementations.
+    if db.trait_function_body(id).ok()?.is_some() {
+        return None;
+    }
+
+    let signature = db.trait_function_signature(id).ok()?;
+
+    let generic_parameters = db.trait_function_generic_params(id).ok()?;
+    let generic_parameters_bracket = if generic_parameters.is_empty() {
+        String::new()
+    } else {
+        let formatted_parameters = generic_parameters
+            .into_iter()
+            .map(|parameter| generic_parameter_code(db, parameter, type_concretizer))
+            .collect::<Option<Vec<_>>>()?
+            .join(", ");
+
+        format!("<{formatted_parameters}>")
+    };
+
+    let parameters = signature
+        .params
+        .iter()
+        .map(|parameter| function_parameter(db, parameter, type_concretizer))
+        .collect::<Option<Vec<_>>>()?
+        .join(", ");
+
+    let name = id.name(db);
+    let title = Some(format!("fn {name}{generic_parameters_bracket}({parameters})"));
+
+    let return_type = type_concretizer.rewrite(signature.return_type).ok()?;
+    let return_type =
+        if return_type.is_unit(db) { None } else { Some(format!("-> {}", return_type.format(db))) };
+
+    let implicits = match &signature.implicits[..] {
+        [] => None,
+        types => Some(format!("implicits({})", types.iter().map(|ty| ty.format(db)).join(", "))),
+    };
+
+    let nopanic = if !signature.panicable { Some(String::from("nopanic")) } else { None };
+
+    let body: Option<String> = Some(String::from("{}"));
+
+    Some([title, return_type, implicits, nopanic, body].into_iter().flatten().join(" "))
+}
+
+/// Formats [`GenericParam`] to be used as a free-standing type parameter
+/// (not belonging to the trait) in a generic function's declaration.
```

Comment:
That's not always true, sometimes you substitute trait generic types with concrete types, no?

Reply:
Indeed, docstring is inaccurate. This function uses substitution rewriter on impl-related parameters but keeps "free" generics unchanged, so it's functionally correct